skyrim/libs/ic/ifs.py

```diff
@@ -0,0 +1,132 @@
+import numpy as np
+import xarray
+import hashlib
+import os
+import boto3
+from botocore import UNSIGNED
+from botocore.client import Config
+from loguru import logger
+from typing import List
+from earth2mip.datasets.era5 import METADATA
+from earth2mip.initial_conditions import base
+
+LOCAL_CACHE = os.getenv("LOCAL_CACHE") or (os.environ["HOME"] + "/.cache/modulus")
+
+def _get_filename(time: datetime.datetime, lead_time: str, resolution: str = '0p25'):
+    if resolution not in {'0p4-beta', '0p25'}:
+        raise Exception('Unknown resolution for IFS')
+    date_format = f"%Y%m%d/%Hz/ifs/{resolution}/oper/%Y%m%d%H%M%S-{lead_time}-oper-fc.grib2"
+    return time.strftime(date_format)
+
+
+def _get_channel(c: str, **kwargs) -> xarray.DataArray:
+    """
+
+    Parameters:
+    -----------
+    c: channel id
+    **kwargs: variables in ecmwf data
+    """
+    # handle 2d inputs
+    if c in kwargs:
+        return kwargs[c]
+    else:
+        varcode, pressure_level = c[0], int(c[1:])
+        return kwargs[varcode].interp(isobaricInhPa=pressure_level)
+
+
+
+def download_cached(path):
+    cached_path = get_cache_file_path(path)
+    if not os.path.exists(cached_path):
+        logger.debug('Downloading IFS initial condition...')
+        s3 = boto3.client('s3', config=Config(signature_version=UNSIGNED))
+        bucket_name = 'ecmwf-forecasts' 
+        file_key = path.replace(f's3://{bucket_name}/', '')
+        s3.download_file(bucket_name, file_key, cached_path)
+        return cached_path
+    return cached_path
+
+
+def get_cache_file_path(path, local_cache_path: str = LOCAL_CACHE):
+    sha = hashlib.sha256(path.encode())
+    filename = sha.hexdigest()
+    try:
+        os.makedirs(local_cache_path, exist_ok=True)
+    except PermissionError as error:
+        logger.error(
+            "Failed to create cache folder, check permissions or set a cache"
+            + " location using the LOCAL_CACHE environment variable"
+        )
+        raise error
+    except OSError as error:
+        logger.error(
+            "Failed to create cache folder, set a cache"
+            + " location using the LOCAL_CACHE environment variable"
+        )
+        raise error
+    return os.path.join(local_cache_path, filename)
+
+
+def get(time: datetime.datetime, channels: List[str]):
+    root = 's3://ecmwf-forecasts/'
+    path = root + _get_filename(time, "0h")
+    local_path = download_cached(path)
+    dataset_0h = xarray.open_dataset(local_path, engine="cfgrib")
+    # get t2m and other things from 12 hour forecast initialized 12 hours before
+    # The HRES is only initialized every 12 hours
+    path = root + _get_filename(time - datetime.timedelta(hours=12), "12h")
+    local_path = download_cached(path)
+    forecast_12h = xarray.open_dataset(local_path, engine="cfgrib")
+    # TODO: some of these channels seem to be omitted. Revisit.
+    channel_data = [
+        _get_channel(
+            c,
+            z=dataset_0h.gh * 9.81,
```

Comment:
we should double-check this, the definition might be different for different models, i.e. without w/o 9.81

Reply:
best to have a channels summary for all models, seems like this is a potential tripping point

Reply:
geopotential is in ERA5 is defined as: https://confluence.ecmwf.int/display/CKB/ERA5%3A+data+documentation
same as IFS: https://confluence.ecmwf.int/display/UER/Geopotential

so converting gh this way should work